src/xdsl/elevate.py

```diff
@@ -0,0 +1,730 @@
+from __future__ import annotations
+from abc import abstractmethod
+from dataclasses import dataclass
+from functools import partial
+from typing import List, Callable
+from xdsl.immutable_ir import *
+from xdsl.pattern_rewriter import *
+
+
+@dataclass
+class IOpReplacement:
+    matched_op: IOp
+    replacement_ops: List[IOp]
+
+
+@dataclass
+class RewriteResult:
+    _result: Union[Strategy, List[IOpReplacement]]
+
+    def flatMapSuccess(self, s: Strategy) -> RewriteResult:
+        if (not self.isSuccess()):
+            return self
+        rr = s.apply(self.result_op)
+        if (not rr.isSuccess()):
+            return rr
+        self += rr
+        return self
+
+    def flatMapFailure(self, f: Callable[[], RewriteResult]) -> RewriteResult:
+        if (not self.isSuccess()):
+            return f()
+        return self
+
+    def __str__(self) -> str:
+        if not self.isSuccess():
+            return "Failure(" + str(self.failed_strategy) + ")"
+        return "Success, " + str(len(self.replacements)) + " replacements"
+
+    def isSuccess(self) -> bool:
+        return isinstance(self._result, List)
+
+    def __iadd__(self, other: RewriteResult):
+        if self.isSuccess() and other.isSuccess():
+            assert self.isSuccess() and other.isSuccess()
+            assert isinstance(self._result, List) and isinstance(
+                other._result, List)
+            self._result += other._result
+            return self
+        raise Exception("invalid concatenation of RewriteResults")
+
+    @property
+    def failed_strategy(self) -> Strategy:
+        assert not self.isSuccess()
+        assert isinstance(self._result, Strategy)
+        return self._result
+
+    @property
+    def replacements(self) -> List[IOpReplacement]:
+        assert self.isSuccess()
+        assert isinstance(self._result, List)
+        return self._result
+
+    @property
+    def result_op(self) -> IOp:
+        assert self.isSuccess()
+        assert not isinstance(self._result, Strategy)
+        return self.replacements[-1].replacement_ops[-1]
+
+
+def success(arg: IOp | List[IOp] | RewriteResult) -> RewriteResult:
```

Comment:
Why do you allow a `RewriteResult` as the argument here?
It seems that this will always result in a runtime exception!?